docs/dev/development/patterns.rst

```diff
@@ -107,6 +107,107 @@ Class                     Method
 ``HTTPServerError`` (5xx) Raise
 ========================= ==================================
 
+Implementing new services
+=========================
+
+Warehouse uses services to provide pluggable functionalities within the codebase. They are implemented using
+`pyramid-service`_. After being registered, services are accessible using the ``find_service`` method of the
+``request`` object.
+
+When adding new services to ``warehouse``, the following checklist serves as a comprehensive guideline to ensure
+you stay on track.
+
+Adding a new service
+~~~~~~~~~~~~~~~~~~~~~
+
+1. Create an Interface for the service. The interface serves as the baseline of the new service (design by
+   contract pattern) and details all methods and attributes shared by the different service implementations.
+
+   Warehouse uses zope_ to define interfaces. The interfaces are usually declared in a file named
+   ``interfaces.py`` within the relevant component, such as ``packaging/interfaces.py``.
+
+2. Create the new service. The service must define all methods and attributes declared in the interface.
+   This implementation contains the core logic of the service features.
```

Comment:
suggestion: services MAY add further methods specific to the service that are not required on all implementations of the interface.